File: apina-core/src/main/java/fi/evident/apina/java/model/JavaModel.java
package fi.evident.apina.java.model;

import fi.evident.apina.java.model.type.JavaBasicType;
import fi.evident.apina.java.model.type.JavaType;

import java.util.*;

import static fi.evident.apina.utils.CollectionUtils.filter;

/**
 * Contains metadata for all loaded classes.
 */
public final class JavaModel {

    private final Map<String,JavaClass> classes = new LinkedHashMap<>();

    public void addClass(JavaClass aClass) {
        JavaClass old = classes.putIfAbsent(aClass.getName(), aClass);
        if (old != null)
            throw new IllegalStateException("Class " + aClass.getName() + " was already added");
    }

    public boolean containsClass(String name) {
        return classes.containsKey(name);
    }

    public Optional<JavaClass> findClass(String name) {
        return Optional.ofNullable(classes.get(name));
    }

    public Optional<JavaClass> findClass(JavaBasicType type) {
        return findClass(type.getName());
    }

    public List<JavaClass> findClassesWithAnnotation(JavaBasicType annotationType) {
        return filter(classes.values(), c -> c.hasAnnotation(annotationType));
    }

    /**
     * Return types for annotations implied by given annotation using same logic as Spring does.
     * E.g. if annotation {@code @Foo} is itself annotated by {@code @Bar}, then annotating an
     * element with {@code @Foo} has the same effect as annotation by {@code @Bar}.
     */
    public Set<JavaBasicType> findAnnotationsImpliedBy(JavaBasicType annotationType) {
        LinkedHashSet<JavaBasicType> result = new LinkedHashSet<>();

        findAnnotationsImpliedBy(annotationType, result);

        return result;
    }

    private void findAnnotationsImpliedBy(JavaBasicType annotationType, Set<JavaBasicType> result) {
        if (result.add(annotationType)) {

            for (JavaClass cl : classes.values())
                if (cl.isAnnotation() && cl.hasAnnotation(annotationType))
                    findAnnotationsImpliedBy(cl.getType().toBasicType(), result);
        }
    }

    public boolean isInstanceOf(JavaType type, Class<?> requiredType) {
        JavaClass javaClass = classes.get(type.getNonGenericClassName());

        if (javaClass != null) {
            return javaClass.getName().equals(requiredType.getName())
                || isInstanceOf(javaClass.getSuperClass(), requiredType)
                || javaClass.getInterfaces().stream().anyMatch(interfaceType -> isInstanceOf(interfaceType, requiredType));

        } else if (type instanceof JavaBasicType) {
            try {
                Class<?> aClass = Class.forName(((JavaBasicType) type).getName());
                return requiredType.isAssignableFrom(aClass);

            } catch (ClassNotFoundException e) {
                return false;
            }
        } else {
            return false;
        }
    }

    public boolean isNumber(JavaBasicType type) {
        return type.isPrimitiveNumber() || isInstanceOf(type, Number.class);
    }

    @Override
    public String toString() {
        return classes.toString();
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/java/model/JavaModel.kt
package fi.evident.apina.java.model

import fi.evident.apina.java.model.type.BoundClass
import fi.evident.apina.java.model.type.JavaType
import fi.evident.apina.java.model.type.TypeEnvironment
import fi.evident.apina.java.reader.ClassDataLoader

/**
 * Contains metadata for all loaded classes.
 */
class JavaModel(private val classDataLoader: ClassDataLoader) {

    fun findClass(name: String): JavaClass? = classDataLoader.loadClass(name)

    fun findClass(type: JavaType.Basic): JavaClass? = findClass(type.name)

    private fun findClasses(namePredicate: (String) -> Boolean): List<JavaClass> =
        classDataLoader.classNames
            .filter { namePredicate(it) }
            .mapNotNull { classDataLoader.loadClass(it) }

    fun findClassesWithAnnotation(namePredicate: (String) -> Boolean, annotationType: JavaType.Basic): List<JavaClass> =
        findClasses(namePredicate)
            .filter { it.hasAnnotation(annotationType) }

    inline fun <reified T : Any> isInstanceOf(type: JavaType) = isInstanceOf(type, T::class.java)

    fun isInstanceOf(type: JavaType, requiredType: Class<*>): Boolean {
        if (type is JavaType.Wildcard)
            return type.upperBound != null && isInstanceOf(type.upperBound, requiredType)

        val javaClass = classDataLoader.loadClass(type.nonGenericClassName)

        return when {
            javaClass != null ->
                javaClass.name == requiredType.name
                        || isInstanceOf(javaClass.superClass, requiredType)
                        || javaClass.interfaces.any { isInstanceOf(it, requiredType) }

            type is JavaType.Basic ->
                try {
                    requiredType.isAssignableFrom(Class.forName(type.name))

                } catch (e: ClassNotFoundException) {
                    false
                }

            else ->
                false
        }
    }

    fun isIntegral(type: JavaType.Basic) =
        type.isPrimitiveIntegral || isInstanceOf<Short>(type) || isInstanceOf<Int>(type) || isInstanceOf<Long>(type)

    fun isNumber(type: JavaType.Basic) = type.isPrimitiveNumber || isInstanceOf<Number>(type)

    fun classesUpwardsFrom(javaClass: BoundClass): List<BoundClass> {
        val result = ArrayList<BoundClass>()

        fun recurse(c: BoundClass) {
            if (result.any { c.javaClass == it.javaClass })
                return

            result += c

            c.javaClass.interfaces
                .asSequence()
                .mapNotNull { boundClassFor(it, c.environment) }
                .forEach(::recurse)
        }

        var cl: BoundClass? = javaClass
        while (cl != null) {
            recurse(cl)
            cl = boundClassFor(cl.javaClass.superClass, cl.environment)
        }

        return result
    }

    private fun boundClassFor(type: JavaType, env: TypeEnvironment): BoundClass? =
        findClass(type.nonGenericClassName)?.let { c ->
            if (type is JavaType.Parameterized)
                BoundClass(c, c.schema.apply(type.arguments.map { it.resolve(env) }))
            else
                BoundClass(c, TypeEnvironment.empty())
        }

    fun findDirectSubclassesInSamePackage(cl: JavaClass): List<JavaClass> {
        val packageName = cl.type.packageName
        return findClasses { JavaType.packageNameForClassName(it) == packageName }
            .filter { it.superClass == cl.type }
    }
}
